DAS-148, porting SmoothData from Mantid to scipp

mads-bertelsen opened the pull request:
Resulting code from DAS-148 where the Mantid function SmoothData was ported to scipp.

SmoothData updates each point to an average of the surrounding points and itself.

The Mantid implementation uses a rolling average, where floating point errors can become problematic, especially for the errors that uses a squareroot which has to use absolute values to avoid imaginary errors.
Meanwhile the scipp version takes the mean of each point set, which is slower but both simpler and has no issues with accumulation of floating point error.

A Jupyter Notebook demo is available that shows the implementations produce equivalent results.

Commit:
Resulting code from DAS-148 where the Mantid function SmoothData was ported to scipp.

SmoothData updates each point to an average of the surrounding points and itself.

The Mantid implementation uses a rolling average, where floating point errors can become problematic, especailly for the errors that ues a squareroot which has to use absolute values to avoid imaginary errors.
Meanwhile the scipp version takes the mean of each point set, which is slower but both simpler and has no issues with accumulation of floating point error.

A Jupyter Notebook demo is available that shows the implementations produce eqvivalent results.

SimonHeybrock commented on smoothdata.py:
```diff
@@ -0,0 +1,52 @@
+import scipp as sc
+
+
+def smoothdata(variable, dim=None, NPoints=3):
```
Consistently use snake case?

mads-bertelsen replied:
Agreed, will update.

SimonHeybrock commented on smoothdata.py:
```diff
@@ -0,0 +1,52 @@
+    if NPoints < 3:
+        raise ValueError("smoothdata needs NPoints of 3 or higher.")
+
+    data_length = len(variable[dim, :].values)
```
I am not sure what the intention here is? Are you trying to get the length in dimensions `dim`? What you are actually getting is the length of the outermost dimension (which should be what you want if the input is 1-D).

Is this what you are looking for:
```python
data_length = dict(zip(variable.dims, variable.shape))[dim]
```
This is a bit cumbersome, we should consider adding a helper to scipp that does this.

mads-bertelsen replied:
I was trying to make sure it would be possible to smooth over one given dim even if several was present. Your solution certainly does what I intended. Will update.

SimonHeybrock commented on smoothdata.py:
```diff
@@ -0,0 +1,52 @@
+    for index in range(data_length - hr, data_length):  # End part
+        start_index = int(index - hr)
+        end_index = data_length  # Always use data length as endpoint
+        out[dim, index] = sc.mean(variable[dim, start_index:end_index], dim)
```
```python
for index in range(data_length):
    begin = max(0, index)
    end = min(data_length, index + 1)
    out[dim, index] = sc.mean(variable[dim, begin, end], dim)
```
?

mads-bertelsen replied:
I did consider using min and max, but it adds unnecessary branching. The code I ended up with should be significantly faster, but I did not benchmark.

SimonHeybrock replied:
`min` and `max`of integers should be extremely fast and completely irrelevant for performance. As far as I am aware compilers do *not* generate branches from `min` and `max` (but even if they did, this is completely irrelvant on this level, have a look at what memory allocation (done by `mean` to store the output) costs compared to this).
You may even gain something because having more code is typically slower.

mads-bertelsen replied:
I timed the two solutions with 50000 points, and they are within 1% of each other as you expected. Will update to the shorter version.

Commit:
Update after review

Response to reviewer comments, updated function and module name to snakecase and used better approach for finding length of chosen dimension.

Commit:
Review response

Consolidated the three loops into one by using min and max for index handling.

mads-bertelsen commented:
I have pushed a new version that addresses your concerns.

SimonHeybrock reviewed (changes requested):
You have some copies of files now.

Commit:
Removed files that should not have been included.
